### PR TITLE
Support non-multiple and non-named values in Yields and Receives

### PR DESCRIPTION
**Pull request by the-13th-letter**
This PR fixes https://github.com/mkdocstrings/griffe/issues/263 by applying the same logic from the `Returns` section parser to the `Yields` and `Receives` parser.  The `Yields` section parser is configured with the same two configuration variables `returns_multiple_items` and `returns_named_value`, whereas the `Receives` section parser is configured via the new `receives_multiple_items` and `receives_named_value` configuration variables. Furthermore, the parsing of multiple or single docstring block contents, of named or unnamed parameters, and the unpacking of generator and/or tuple return annotations has been refactored into separate functions.

---

<b>Contents:</b>

- 8a78e68e7702d5b874daa21aeca58943eb9cf7d6 duplicates the non-multiple return and the named value parsing support for the `Yields` and `Receives` section parsers, essentially via copy-and-paste. Includes suitable tests, based on the `Returns` section tests.
- 756118fc0da4866c80fde0b456cbb497124c0a18 implements the aforementioned refactored functions. This commit is functionally optional, but hopefully useful with respect to future ease of maintenance.
- e01447a5980ca2c6e94c93f8df7618deb59d627a fixes linting errors.
- 8e7cb30fac0df5e619da94b102d5c88de0b36a18 adds documentation for the above features.

### CODE REVIEW

**Comment by the-13th-letter** [comment]
Concerning code style, I come from a `black`/`ruff` environment, and found no settings in your code repository. So my formatting and linting may be off.

File: src/_griffe/docstrings/google.py [hunk 1]
```diff
@@ -491,22 +548,16 @@ def _read_returns_section(
...
+                orig_annotation = annotation
+                annotation = _maybe_destructure_annotation(
+                    annotation, index, multiple=returns_multiple_items
+                )
+                if annotation == orig_annotation:
+                    annotation = _maybe_destructure_annotation(
+                        _unpack_generators(annotation, 2),
+                        index,
+                        multiple=returns_multiple_items,
+                    )
```
Comment (pawamoy) [inline src/_griffe/docstrings/google.py]: It all looks good, except maybe this part. I find it a bit hard to follow, compared to the previous version.

Since `_unpack_generators` is called without `mandatory=True` here, couldn't we directly do this?

```python
                annotation = _maybe_destructure_annotation(
                    _unpack_generators(annotation, 2),
                    index,
                    multiple=returns_multiple_items,
                )
```

I'll try this locally and see if tests pass.

**Review by pawamoy — COMMENTED** [review]
Thanks a lot! I appreciate the effort on splitting commits to ease review :slightly_smiling_face:

File: src/_griffe/docstrings/google.py [hunk 2]
```diff
@@ -491,22 +552,18 @@ def _read_returns_section(
...
+                    annotation,
+                    index,
+                    multiple=returns_multiple_items,
+                )
+                if annotation == orig_annotation:
+                    annotation = _maybe_destructure_annotation(
+                        _unpack_generators(annotation, 2),
+                        index,
+                        multiple=returns_multiple_items,
+                    )
```
Comment (pawamoy) [inline src/_griffe/docstrings/google.py:566]: Tested locally and tests pass with this change:

```suggestion
                annotation = _maybe_destructure_annotation(
                    _unpack_generators(annotation, 2),
                    index,
                    multiple=returns_multiple_items,
                )
```

I wonder if we're missing something though, but IIRC tests are quite comprehensive?

File: src/_griffe/docstrings/google.py [hunk 3]
```diff
@@ -442,6 +442,76 @@ def _read_warns_section(
...
+
+def _unpack_generators(
+    annotation: Any,
+    generator_pos: int,
+    *,
+    mandatory: bool = False,
+) -> Any:
+    if annotation.is_generator:
+        return annotation.slice.elements[generator_pos]
+    if annotation.is_iterator and generator_pos in {0, 2}:
```
Comment (pawamoy) [inline src/_griffe/docstrings/google.py:495]: I'm not entirely convinced about the `generator_pos` check (`generator_pos in {0, 2}`). Raising "must be a generator" because `generator_pos` is > 2 would feel wrong, and we only ever pass 0, 1 and 2 anyway, so maybe we could just drop this check?

File: src/_griffe/docstrings/google.py (hunk 2 above)
Comment (the-13th-letter) [reply src/_griffe/docstrings/google.py:566]: It's your call. I translated the code very directly from the non-PR version, and it included this try-unpacking-else-un-generator-then-unpack dance. I *did* get a "organically grown and never pruned" vibe from this part, but I didn't study Griffe's workings and code history deeply enough to be comfortable with pruning this section… at least not without double-checking with you first.

In any case, I don't see any problems with your suggestion, based solely on reading and rewriting that code section.

File: src/_griffe/docstrings/google.py (hunk 3 above)
Comment (the-13th-letter) [reply src/_griffe/docstrings/google.py:495]: Looking back at this, I think I should have set up an enum instead, instead of magic numbers. Makes for a cleaner interface, and better type checking support.

File: src/_griffe/docstrings/google.py (hunk 2 above)
Comment (pawamoy) [reply src/_griffe/docstrings/google.py:566]: Thanks, that makes sense.

> an "organically grown and never pruned" vibe

Definitely. The parsers were never something I felt interested in (for many reasons), and I just coded them lazily, in an imperative mood. I didn't spend much time trying to make them elegant or efficient, as long as they are working as intended. Our tests suite says the suggestion works, so lets apply it. If it turns out the tests are insufficient, we'll fix that then :relaxed:

File: src/_griffe/docstrings/google.py (hunk 3 above)
Comment (the-13th-letter) [reply src/_griffe/docstrings/google.py:495]: Not converting the section to enums, [as requested](https://github.com/mkdocstrings/griffe/pull/322#pullrequestreview-2289652118).

**Comment by the-13th-letter** [comment]
GitHub doesn't seem to recognize a PR that has been reset to an earlier, already-reviewed state as already reviewed…

Also, there's an apparently unrelated, failing CI test on Py3.13/Ubuntu-lowest-direct that's been haunting this PR since its inception.

**Review by pawamoy — APPROVED** [review]

**Comment by pawamoy** [comment]
Don't mind the 3.13-lowest failure, it's unrelated indeed and it's allowed to fail.

**Comment by pawamoy** [comment]
> Concerning code style, I come from a black/ruff environment, and found no settings in your code repository. So my formatting and linting may be off.

Just saw this, sorry. There's a `make vscode` command you can use to configure VSCode for the project :slightly_smiling_face: It overrides local files.

**Comment by pawamoy** [comment]
Thanks again!

**Comment by pawamoy** [comment]
I refactored the logic again in https://github.com/mkdocstrings/griffe/commit/a80bd3c0cc14e5f6efc30fb804b8c7fccb319276, I think it's even more readable now.

**Comment by the-13th-letter** [comment]
That's fine. I wasn't sure if I should touch the "where is the annotation looked up in?" part, so my version doesn't quite reap the same readability benefits as yours.
